### PR TITLE
[jit] Add immutable dict support

### PR DESCRIPTION
**Pull request by driazati**
This PR adds basic support (creation and indexing) for immutable dictionaries in Script. This includes Python/string frontend support and a `IValue::GenericDict` type backed by a `std::unordered_map`. Only `str`, `int`, and `float` are supported as keys, any type can be a value. Structure is pretty similar to list.

### CODE REVIEW

File: aten/src/ATen/core/ivalue.h [hunk 1]
```diff
@@ -166,6 +204,22 @@ struct CAFFE2_API IValue final {
     std::swap(tag, rhs.tag);
   }
 
+
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.h]: Let's avoid declaring equality for IValues generally and instead have a definition used by Dict objects specifically.  The problem is that equality for these values is poorly defined (is it python ==, is, or something else?). It is better to just ask how we want to dict to behave.

File: aten/src/ATen/core/ivalue.h [hunk 2]
```diff
@@ -811,6 +891,17 @@ inline bool IValue::isSameIdentity(IValue& rhs) {
         && this->payload.as_intrusive_ptr == rhs.payload.as_intrusive_ptr;
   }
 }
-
-
 } // namespace c10
+
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.h]: Same for hashing: it would be better to not expose this on IValue's public API.

Both changes can be done by specifying additional template arguments to the unordered_map.

File: torch/csrc/jit/ir.cpp [hunk 3]
```diff
@@ -1218,6 +1218,30 @@ Node* Graph::createListUnpack(Value* v, size_t size) {
   return n;
 }
 
+Node* Graph::createDict(
+    const TypePtr& key_type,
+    const TypePtr& value_type,
+    std::unordered_map<Value*, Value*>& values) {
+
+  auto n = create(prim::DictConstruct, 1);
+  for (auto v : values) {
```
Comment (zdevito) [inline torch/csrc/jit/ir.cpp]: nondeterministic IR warning! Iterating over unordered map will cause the IR to be different every time.  There is no need for the API to take a map, it is just being used as a list here anyway.

File: torch/csrc/jit/ir.cpp [hunk 4]
```diff
@@ -1218,6 +1218,30 @@ Node* Graph::createListUnpack(Value* v, size_t size) {
...
+    n->addInput(v.first) ;
+    n->addInput(v.second);
+  }
+  n->output()->setType(DictType::create(key_type, value_type));
+  return n;
+}
+
+Node* Graph::createDictIndex(Value* dict, Value* index) {
+  auto n = create(prim::DictIndex, {dict, index});
+  auto dict_type = dict->type()->expect<DictType>();
```
Comment (zdevito) [inline torch/csrc/jit/ir.cpp]: It's a good idea to assert the correct index type here as well.

File: torch/csrc/jit/passes/python_print.cpp [hunk 5]
```diff
@@ -776,6 +795,15 @@ struct PythonPrintPass {
           printValueList(stmt, node->inputs(), "[", "]");
         }
       } break;
+      case prim::DictConstruct: {
+        stmt << "annotate(" << node->output()->type()->python_str() << ", ";
```
Comment (zdevito) [inline torch/csrc/jit/passes/python_print.cpp]: Minor nit: for lists we do not annotate if it is not necessary. For instance if the list is non-empty, or if it is empty but the type is `List[Tensor]`. Is something like that possible here? It will make the outputted form more readable and potentially more robust to use changing things around.

File: torch/csrc/jit/pybind_utils.h [hunk 6]
```diff
@@ -187,6 +195,20 @@ inline IValue toIValue(
           return createGenericList(obj, elem_type);
       }
     }
+    case TypeKind::DictType: {
+      const auto& dict_type = type->expect<DictType>();
+      switch (dict_type->getKeyType()->kind()) {
+        case TypeKind::IntType:
```
Comment (zdevito) [inline torch/csrc/jit/pybind_utils.h]: We should centralize the logic for what is a valid key should be somewhere, e.g. in ivalue. Otherwise when it gets expanded, someone will forget to modify all these error checks.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 7]
```diff
@@ -789,6 +789,40 @@ RegisterOperators reg({
...
+    Operator(
+        prim::DictIndex,
+        [](const Node* node) {
+          return [=](Stack& stack) {
+            auto index = pop(stack);
+            auto dict = pop(stack).toGenericDict();
+            const auto& elems = dict->elements();
+            auto value = elems.find(index);
+            if (value == elems.end()) {
+              AT_ERROR("DictIndex could not find key '", index, "'");
```
Comment (zdevito) [inline torch/csrc/jit/register_prim_ops.cpp]: can you make the error message match pythons as close as reasonable?

File: torch/csrc/jit/script/compiler.cpp [hunk 8]
```diff
@@ -2241,8 +2241,39 @@ struct to_ir {
...
+
+        if (type_hint && type_hint->kind() == TypeKind::DictType) {
+          auto dict_type = type_hint->expect<DictType>();
+          key_type = dict_type->getKeyType();
+          value_type = dict_type->getValueType();
+        } else if (!values.empty()) {
+          key_type = keys.at(0)->type();
+          value_type = values.at(0)->type();
+        } else {
+          throw ErrorReport(tree) << "Cannot get key and value types without a"
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: For lists, we just assume tensor. For dict, we could consider assuming string -> tensor. I am not really sure how people will typically use dicts, however.

File: torch/csrc/jit/script/compiler.cpp [hunk 9]
```diff
@@ -2241,8 +2241,39 @@ struct to_ir {
...
+        }
+
+        JIT_ASSERT(key_type != nullptr && value_type != nullptr);
+
+        std::unordered_map<Value*, Value*> value_ptrs;
+        for (size_t i = 0; i < keys.size(); ++i) {
+          value_ptrs[keys[i]] = values[i];
+        }
+
+        return graph->insertNode(graph->createDict(key_type, value_type, value_ptrs))->output();
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: Why not just make createDict take keys and values as separate lists? That is already what was created here.

File: torch/csrc/jit/script/compiler.cpp [hunk 10]
```diff
@@ -2241,8 +2241,39 @@ struct to_ir {
         auto values = getValues(ll.inputs(), /*maybe_unpack=*/true);
         return graph->insertNode(graph->createTuple(values))->output();
       } break;
+      case TK_DICT_LITERAL: {
+        auto dl = DictLiteral(tree);
+        auto keys = getValues(dl.key_inputs(), /*maybe_unpack=*/true);
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: Evaluation order warning: if I write
`{ print('a'): print('b'), print('c'), print('d') }`. Then it should print `a b c d` but this will print `a c b d`.

File: torch/csrc/jit/script/compiler.cpp [hunk 11]
```diff
@@ -2528,9 +2569,12 @@ struct to_ir {
     } else if (auto tuple_type = gatherable->type()->cast<TupleType>()) {
       auto* idx = emitExpr(subscript_exprs[0]);
       return emitTupleIndex(loc, gatherable, idx);
+    } else if (auto dict_type = gatherable->type()->cast<DictType>()) {
+      auto* idx = emitExpr(subscript_exprs[0]);
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: There is no type checking here, it will accept any argument as a key!

**Review by zdevito — COMMENTED** [review]
This looks good! The overall structure works, I just have some minor places where I think there are bugs.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 12]
```diff
@@ -789,6 +789,40 @@ RegisterOperators reg({
             };
           }
         }),
+    Operator(
+      prim::DictConstruct,
```
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: You will need to add a case for prim::DictConstruct & prim::DictIndex in alias_analysis

**Review by eellison — COMMENTED** [review]
I haven't looked really looked into this but I see that it is missing alias analysis. 

Look at how it's done for ListType - you will need to add a map for dictTypeAliases, add DictType to shouldAnnotate(), & add cases for both prim ops you introduced

**Comment by sidazhang** [comment]
So this can be used as input. Can this be used as output as well?

For example

def forward():
    return {“some_key”: tensor}

**Comment by driazati** [comment]
@sidazhang that should be fine, something like this works:

```python
@torch.jit.script
def create_dict():
    # type: () -> Dict[str, Tensor]
    return {'word1': torch.ones(1) + 2, 'word2': torch.ones(1) + 3}
```

**Comment by sidazhang** [comment]
I am not familiar with the codebase so I could just be talking rubbish.

Can you also support List[Dict]? And if so, do you need to change this function too?

```
TypePtr incompleteInferTypeFrom(const IValue& value) {
  if (value.isTensor()) {
    return CompleteTensorType::create(value.toTensor());
  } else if (value.isDouble()) {
    return FloatType::get();
  } else if (value.isInt()) {
    return IntType::get();
  } else if (value.isBool()) {
    return BoolType::get();
  } else if (value.isString()) {
    return StringType::get();
  } else if (value.isIntList()) {
    return ListType::ofInts();
  } else if (value.isTensorList()) {
    return ListType::ofTensors();
  } else if (value.isBoolList()) {
    return ListType::ofBools();
  } else if (value.isDoubleList()) {
    return ListType::ofFloats();
  } else if (value.isTuple()) {
    return TupleType::create(fmap(value.toTuple()->elements(), incompleteInferTypeFrom));
  } else if (value.isDevice()) {
    return DeviceObjType::get();
  }
  AT_ERROR("Type cannot be accurately recovered from this IValue.");
}
```

File: torch/csrc/jit/register_aten_ops.cpp [hunk 13]
```diff
@@ -0,0 +1,1217 @@
+#include <torch/csrc/autograd/profiler.h>
```
Comment (eellison) [inline torch/csrc/jit/register_aten_ops.cpp]: Why was this file changed ? I don't see anything related to Dict here

**Comment by rkaplan** [comment]
This is awesome. "any type can be a value" -> does this mean nested dictionaries are supported? Would be amazing if so.

**Comment by zdevito** [comment]
@rkaplan yes, we plan to support nested dictionaries.

File: aten/src/ATen/core/ivalue.h [hunk 14]
```diff
@@ -65,6 +65,44 @@ struct CAFFE2_API List : c10::intrusive_ptr_target {
...
+};
+
+struct DictEqualTo {
+  bool operator()(const IValue& lhs, const IValue& rhs) const;
+};
+template <typename Key, typename Value>
+using DictUnorderedMap =
+  std::unordered_map<Key, Value, DictHash, DictEqualTo>;
+
+  template <typename Key, typename Value>
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.h]: DictHash and DictEqualTo make this only work with IValues. Why not just write a non-templated Dict? No reason for complexity that isn't useful.

File: aten/src/ATen/core/jit_type.h [hunk 15]
```diff
@@ -510,6 +511,94 @@ struct CAFFE2_API ListType : public SingleElementType<TypeKind::ListType, ListTy
...
+        AT_ERROR(
+            "Cannot create dict for key type ",
+            key->str(),
+            ", only int, float, and string keys are supported");
+    }
+  }
+
+  std::string str() const override {
+    std::stringstream ss;
+    ss << getKeyType()->str() << "[" << getValueType()->str() << "]";
```
Comment (zdevito) [inline aten/src/ATen/core/jit_type.h]: Not sure it makes sense to have a different syntax, 'just because' for dictionaries. It is so close to array syntax `a[4] vs a[int]` that it will be confusing.

File: torch/csrc/jit/script/compiler.cpp [hunk 16]
```diff
@@ -2241,8 +2273,35 @@ struct to_ir {
         auto values = getValues(ll.inputs(), /*maybe_unpack=*/true);
         return graph->insertNode(graph->createTuple(values))->output();
       } break;
+      case TK_DICT_LITERAL: {
+        auto dl = DictLiteral(tree);
+        std::vector<Value*> keys, values;
+        std::tie(keys, values) = getValues(
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: All of the above stuff above dealing with a paired list of named values seems like huge overkill compared to:

```
auto keys = dl.key_inputs().tree()->trees();
auto values = dl.value_inputs().tree()->trees();
for(size_t i = 0; i < keys.size(); ++i) {
  keys.push_back(emitExpr(keys[i]));
  values.push_back(emitExpr(values[i]));
}
```

**Review by zdevito — APPROVED** [review]
Looks good.  Just a few nits where things can be simplified.

**Comment by sidazhang** [comment]
This is like a Christmas present :)

❤️
